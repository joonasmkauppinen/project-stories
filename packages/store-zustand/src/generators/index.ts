export * from './generateCard/generateCard';
export * from './generateTextLayer/generateTextLayer';
export * from './generateImageLayer/generateImageLayer';
