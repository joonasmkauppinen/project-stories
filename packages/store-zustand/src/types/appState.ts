export type CardState = 'idle' | 'hovered' | 'active';

export type LayerState = 'idle' | 'hovered' | 'active';

export interface LayerMetaState {
  parentCardActive: boolean;
}

export type ID = string;

export type LayerAnimationType =
  | 'drop'
  | 'fade-in'
  | 'fly-in-bottom'
  | 'fly-in-left'
  | 'fly-in-right'
  | 'fly-in-top'
  | 'pan-down'
  | 'pan-left'
  | 'pan-right'
  | 'pan-up'
  | 'pulse'
  | 'rotate-in-left'
  | 'rotate-in-right'
  | 'twirl-in'
  | 'whoosh-in-left'
  | 'whoosh-in-right'
  | 'zoom-in'
  | 'zoom-out';

export type MimeType =
  | 'audio/mpeg'
  | 'image/gif'
  | 'image/jpeg'
  | 'image/png'
  | 'image/svg+xml'
  | 'text/plain'
  | 'video/mp4';

export type LayerType = 'text' | 'image' | 'video' | 'audio' | 'svg';

export interface Coordinate {
  x: number;
  y: number;
}

export interface Size {
  width: number;
  height: number;
}

export interface LayerResource {
  alt: string;
  height: number;
  id: ID;
  mimeType: MimeType;
  src: string;
  type: LayerType;
  width: number;
}

export interface BaseLayer {
  animation?: {
    type: LayerAnimationType;
    delay?: number;
    duration?: number;
  };
  sortOrderIndex: number;
  position: Coordinate;
  screenPosition: Coordinate;
  size: Size;
  state: LayerState;
  metaState: LayerMetaState;
  name: string;
  type?: LayerType;
}

export interface TextLayer extends BaseLayer {
  type: 'text';
  value: string;
}

export type Layer = TextLayer;

export interface Layers {
  [key: string]: Layer;
}

export interface Card {
  sortOrderIndex: number;
  screenPosition: Coordinate;
  autoAdvance: boolean;
  duration: number;
  layers: Layers;
  name: string;
  state: CardState;
}

export interface Cards {
  [key: string]: Card;
}

// TODO: refactor to SelectedLayers and SelectedCards
export interface SelectionItem {
  id: ID;
  parentId?: ID;
}

export interface SelectedLayer {
  cardId: ID;
  layerId: ID;
}

export interface SelectedCard {
  cardId: ID;
}

export type StoryboardTool = 'move' | 'text' | 'hand';

export interface UserInteraction {
  isDragging: boolean;
  isEditingText: boolean;
}

export interface AppState {
  cards: Cards;
  currentTool: StoryboardTool;
  selectedLayers: SelectedLayer[];
  selectedCards: SelectedCard[];
  userInteraction: UserInteraction;
}
