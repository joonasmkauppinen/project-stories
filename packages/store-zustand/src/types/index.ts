export * from './appState';
export * from './layerActions';
export * from './testUtils';
