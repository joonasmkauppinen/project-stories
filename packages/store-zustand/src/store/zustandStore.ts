import create from 'zustand';
import { devtools } from 'zustand/middleware';

import { AppState, LayerActions } from '../types';
import {
  addNewCard,
  addTextLayerToCard,
  deleteSelectedCards,
  deleteSelectedLayers,
  deselectAll,
  onDragSelection,
  setCardStateToActive,
  setCardStateToHovered,
  setCardStateToIdle,
  setIsDraggingToFalse,
  setIsDraggingToTrue,
  setLayerStateToActive,
  setLayerStateToHovered,
  setLayerStateToIdle,
  setToolToMove,
  setToolToText,
  updateElementScreenPosition,
} from '../actions';
import { generateCard } from '../generators';

const [cardId, emptyCard] = generateCard({ sortOrderIndex: 0 }).idWithData;

export const initialState: AppState = {
  cards: {
    [cardId]: emptyCard,
  },
  currentTool: 'move',
  selectedCards: [],
  selectedLayers: [],
  isDragging: false,
};

export const useStore = create<AppState>(devtools((_) => initialState));

export const selectCards = (state: AppState) => state.cards;
export const selectCurrentTool = (state: AppState) => state.currentTool;
export const selectIsDragging = (state: AppState) => state.isDragging;
export const selectSelectedCards = (state: AppState) => state.selectedCards;
export const selectSelectedLayers = (state: AppState) => state.selectedLayers;

export const actions: LayerActions = {
  addNewCard,
  addTextLayerToCard,
  deleteSelectedLayers,
  deleteSelectedCards,
  deselectAll,
  onDragSelection,
  setLayerStateToActive,
  setLayerStateToHovered,
  setLayerStateToIdle,
  setCardStateToIdle,
  setCardStateToActive,
  setCardStateToHovered,
  setIsDraggingToFalse,
  setIsDraggingToTrue,
  setToolToMove,
  setToolToText,
  updateElementScreenPosition,
};
