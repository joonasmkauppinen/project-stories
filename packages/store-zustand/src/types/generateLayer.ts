import { ID } from './appState';

export interface GenerateLayerReturnType<LayerType> {
  layerId: ID;
  layerData: LayerType;
  idWithData: [string, LayerType];
}
