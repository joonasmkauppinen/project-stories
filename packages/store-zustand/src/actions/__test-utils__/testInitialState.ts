import { AppState } from '../../types/appState';
import { initialState } from '../../store/zustandStore';

/**
 * Same as the real `initialState`, but the `cards`object is set to empty
 * object.
 */
export const testInitialState: AppState = {
  ...initialState,
  cards: {},
};
