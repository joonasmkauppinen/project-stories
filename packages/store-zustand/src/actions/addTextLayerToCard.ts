import produce from 'immer';
import { generateLayer } from '../generators';

import { useStore } from '../store/zustandStore';
import { AppState, ID } from '../types/appState';

export interface AddTextLayerToCardPayload {
  cardId: ID;
  top: number;
  left?: number;
  width?: number;
  height?: number;
  value?: string;
}

export type AddTextLayerToCard = (payload: AddTextLayerToCardPayload) => void;

export const addTextLayerToCard: AddTextLayerToCard = ({
  cardId,
  top,
  height,
  left,
  value,
  width,
}) =>
  useStore.setState(
    produce<AppState>((draft) => {
      const state = useStore.getState();
      const sortOrderIndex = Object.keys(state.cards[cardId].layers).length;
      const { layerId, layerData } = generateLayer({
        sortOrderIndex,
        type: 'text',
        top,
        left,
        width,
        height,
        value,
      });

      // TODO: add clear selections helper function.
      draft.selection = [];

      draft.cards[cardId].layers[layerId] = layerData;
      draft.currentTool = 'move';
    })
  );
