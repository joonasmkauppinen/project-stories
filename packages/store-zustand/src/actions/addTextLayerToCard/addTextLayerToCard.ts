import produce from 'immer';

import { AppState, ID, TestTextLayerOverrides } from '../../types';
import { generateTextLayer } from '../../generators';
import { useStore } from '../../store/zustandStore';

export interface AddTextLayerToCardPayload {
  cardId: ID;
  top: number;
  left?: number;
  width?: number;
  height?: number;
  value?: string;
  testOverrides?: TestTextLayerOverrides;
}

export type AddTextLayerToCard = (payload: AddTextLayerToCardPayload) => void;

export const addTextLayerToCard: AddTextLayerToCard = ({
  cardId,
  top,
  height,
  left,
  value,
  width,
  testOverrides,
}) =>
  useStore.setState(
    produce<AppState>((draft) => {
      const state = useStore.getState();

      // TODO: add clear selections helper function.
      state.selectedLayers.forEach(({ cardId, layerId }) => {
        draft.cards[cardId].layers[layerId].state = 'idle';
      });
      draft.selectedLayers = [];

      state.selectedCards.forEach(({ cardId }) => {
        draft.cards[cardId].state = 'idle';
      });
      draft.selectedCards = [];

      const sortOrderIndex = Object.keys(state.cards[cardId].layers).length;
      const { layerId, layerData } = generateTextLayer({
        sortOrderIndex,
        top,
        left,
        width,
        height,
        value,
        testOverrides,
      });

      draft.cards[cardId].layers[layerId] = layerData;
      draft.currentTool = 'move';
    })
  );
