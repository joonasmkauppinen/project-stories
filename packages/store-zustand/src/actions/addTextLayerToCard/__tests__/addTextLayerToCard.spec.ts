import { AppState, TestCardId, TestLayerId } from '../../../types';
import { generateCard, generateLayer } from '../../../generators';
import { useStore } from '../../../store/zustandStore';

import { testInitialState } from '../../__test-utils__/testInitialState';

import { addTextLayerToCard } from '../addTextLayerToCard';

describe('Action - addTextLayerToCard', () => {
  test('Adds new layer to the card correctly', () => {
    const cardId: TestCardId = 'test_card_id_0';
    const emptyCard = generateCard({
      sortOrderIndex: 0,
      testOverrides: { id: cardId },
    }).idWithData;

    const initialStateWithEmptyCard: AppState = {
      ...testInitialState,
      cards: {
        ...emptyCard,
      },
    };

    useStore.setState(() => initialStateWithEmptyCard);

    const cardLayers = () =>
      Object.values(useStore.getState().cards[cardId].layers);

    expect(cardLayers().length).toEqual(0);

    addTextLayerToCard({
      cardId,
      top: 0,
      testOverrides: { id: 'test_layer_id_0' },
    });

    expect(cardLayers().length).toEqual(1);
    expect(useStore.getState()).toMatchSnapshot();
  });

  test('Clears possible current selected layers correctly', () => {
    const layerId: TestLayerId = 'test_layer_id_0';
    const layerInActiveState = generateLayer({
      sortOrderIndex: 0,
      top: 0,
      type: 'text',
      testOverrides: { id: layerId, properties: { state: 'active' } },
    }).idWithData;

    const cardId: TestCardId = 'test_card_id_0';
    const cardWithOneActiveLayer = generateCard({
      sortOrderIndex: 0,
      layers: { ...layerInActiveState },
      testOverrides: { id: cardId },
    }).idWithData;

    const initialState: AppState = {
      ...testInitialState,
      cards: {
        ...cardWithOneActiveLayer,
      },
      selectedLayers: [{ cardId, layerId }],
    };

    useStore.setState(() => initialState);

    expect(useStore.getState().selectedLayers.length).toEqual(1);
    expect(useStore.getState().selectedLayers).toMatchInlineSnapshot(`
      Array [
        Object {
          "cardId": "test_card_id_0",
          "layerId": "test_layer_id_0",
        },
      ]
    `);

    const newLayerId: TestLayerId = 'test_layer_id_1';
    addTextLayerToCard({ cardId, top: 100, testOverrides: { id: newLayerId } });

    expect(useStore.getState().selectedLayers.length).toEqual(0);
  });

  test('Clears possible current selected cards correctly', () => {
    const cardId: TestCardId = 'test_card_id_0';
    const cardInActiveState = generateCard({
      sortOrderIndex: 0,
      testOverrides: { id: cardId, properties: { state: 'active' } },
    }).idWithData;

    const initialState: AppState = {
      ...testInitialState,
      cards: {
        ...cardInActiveState,
      },
      selectedCards: [{ cardId }],
    };

    useStore.setState(() => initialState);

    expect(useStore.getState().selectedCards.length).toEqual(1);
    expect(useStore.getState().selectedCards).toMatchInlineSnapshot(`
      Array [
        Object {
          "cardId": "test_card_id_0",
        },
      ]
    `);

    addTextLayerToCard({
      cardId,
      top: 0,
      testOverrides: { id: 'test_layer_id_0' },
    });

    expect(useStore.getState().selectedCards.length).toEqual(0);
  });
});
