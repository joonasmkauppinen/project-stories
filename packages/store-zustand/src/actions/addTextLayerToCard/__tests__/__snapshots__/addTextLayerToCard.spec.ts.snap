// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action - addTextLayerToCard Adds new layer to the card correctly 1`] = `
Object {
  "cards": Object {
    "test_card_id_0": Object {
      "autoAdvance": false,
      "duration": 0,
      "layers": Object {
        "test_layer_id_0": Object {
          "metaState": Object {
            "parentCardActive": false,
          },
          "name": "text 1",
          "position": Object {
            "x": 20,
            "y": 0,
          },
          "screenPosition": Object {
            "x": NaN,
            "y": NaN,
          },
          "size": Object {
            "height": 40,
            "width": 320,
          },
          "sortOrderIndex": 0,
          "state": "idle",
          "type": "text",
          "value": "Hello World",
        },
      },
      "name": "Card 1",
      "screenPosition": Object {
        "x": NaN,
        "y": NaN,
      },
      "sortOrderIndex": 0,
      "state": "idle",
    },
  },
  "currentTool": "move",
  "isDragging": false,
  "selectedCards": Array [],
  "selectedLayers": Array [],
}
`;
